src/model/TurnStackStats.java

```diff
@@ -0,0 +1,38 @@
+package model;
+
+import java.util.Stack;
+
+/*
+ * Keeps track of the state of each turn allowing saves and redos.
+ */
+public class TurnStackStats {
+	
+    private Stack<byte[]> stack;
+    private static TurnStackStats turnStackStats = null;
+
+    public TurnStackStats() {
+    	
+        stack = new Stack<byte[]>();
+        if (TurnStackStats.turnStackStats != null) {
+            TurnStackStats.turnStackStats = this;
+            System.out.println("returned turnStack model");
+        } else {
+            TurnStackStats.turnStackStats = this;
+            stack = new Stack<byte[]>();
+        }
+    }
+
+    public void saveTurn(byte[] b) {
+        stack.push(b);
+    }
+
+    public byte[] undoTurn() {
+        if (stack.size() > 1) return stack.pop();
+        else if (stack.size() == 1) return stack.peek();

```

Comment:
this is a mistake on my part, it should be if stack.size() > 0 then pop and if theres nothing then do nothing, except i was returning a byte[] so i was afraid that it was going to error out and insetead of having an empty board, do something weird. we can change in future